### DISCUSSION TITLE
Piecemeal migration v6-v7

### DISCUSSION THREAD

**Discussion by milesingrams** [mantinedev/mantine#4659] [Q&A]
We have a pretty extensive codebase using Mantine and will eventually need to migrate from v6-v7. There is no way we can do this in one giant PR, however the removal of createStyles, styles, and sx invalidates our entire styling system. Is there any way that v7 can be backwards compatible with these APIs and v8 removes them? This will allow anyone using Mantine for multi-page projects to migrate in a piece-by-piece fashion.

Anyone else have thoughts on this or how to execute a migration without doing it all at once?

My suggestion might be to do a migration version in v6 that exposes the css variables. That way you can do the migration slowly in v6 and then jump to v7.

**Comment by rtivital** [comment]
> Is there any way that v7 can be backwards compatible with these APIs?

No, it is not possible, staring from v7 Mantine no longer comes with emotion. It will be difficult to maintain both ways of styling.

> and v8 removes them

What's the difference between removing these in v7 and v8?

> Anyone else have thoughts on this or how to execute a migration without doing it all at once?

Why do you need to migrate? You can continue using v6 as long as you are comfortable with it.

**Reply by milesingrams** [reply]
Migration would mostly be to have access to fixes and improvements made to the library over time. I imagine you will not be adding all new components and fixes to v6 as well as v7? Unless I am incorrect about this assumption. 

I do like the CSS only approach and I think its a good decision if you also allow for migration. If you add the CSS variables generated from the theme to a minor version in v6 then it should be possible for anyone who wants to migrate to do it effectively even in large codebases.

Generally from what I've seen in component libraries and other large API landscapes a common approach to changing functionality is to have a version in which both old and new APIs are supported and then a subsequent version which removes the old deprecated APIs. This makes the transition much less jarring.

**Reply by milesingrams** [reply]
Ahhh I notice you provide CSS variables already: https://mantine.dev/theming/mantine-provider/#css-variables

I thought they were only in v7. I will look into using these as a basis to migrate!

**Reply by viveknigam3003** [reply]
@rtivital Any thoughts on this first line by @milesingrams ?
> Migration would mostly be to have access to fixes and improvements made to the library over time. I imagine you will not be adding all new components and fixes to v6 as well as v7? Unless I am incorrect about this assumption.

I remember we started using v5 in our production codebase, but needed some date-time related components which were introduced in v6 and had to do a migration. But v7 looks extremely different from v6, but definitely future versions of Mantine would, I assume, be built without `emotion` itself. You said we can keep using v6, but I suspect that future versions will have maybe more components which will not support v6 and we might need to upgrade which will become difficult. Any thoughts on how to perform this migration from v6 to v7?

**Reply by rtivital** [reply]
I'm not planning to release 6.x versions. If you need new features, update to the newest version.

> Any thoughts on how to perform this migration from v6 to v7?

I do not know your case. I've updated 4 applications to v7 so far, I had no issues.

**Comment by sridhar-uno** [comment]
Piecemeal migration can be done by replacing createStyles with css modules and sx with styles or again css modules. This will make up bulk of the broken changes. Then you can upgrade to v7.

**Comment by FelixBlaisThon** [comment]
It took us a while to migrate our whole app, but it's worth it. Combobox is a gigachad component
